packages/jupyter-ai/jupyter_ai/actors.py

```diff
@@ -0,0 +1,185 @@
+from enum import Enum
+import os
+import time
+from uuid import uuid4
+from jupyter_ai.models import AgentChatMessage, ChatMessage, HumanChatMessage
+from jupyter_ai_magics.providers import ChatOpenAINewProvider
+from langchain import ConversationChain, OpenAI
+from langchain.chains import ConversationalRetrievalChain
+import ray
+from ray.util.queue import Queue
+from langchain.memory import ConversationBufferMemory
+from langchain.prompts import (
+    ChatPromptTemplate, 
+    MessagesPlaceholder, 
+    SystemMessagePromptTemplate, 
+    HumanMessagePromptTemplate
+)
+from langchain.vectorstores import FAISS
+from langchain.embeddings.openai import OpenAIEmbeddings
+from langchain.document_loaders import DirectoryLoader, TextLoader
+from langchain.text_splitter import RecursiveCharacterTextSplitter
+from jupyter_core.paths import jupyter_data_dir
+
+class ACTOR_TYPE(str, Enum):
+    DEFAULT = "default"
+    FILESYSTEM = "filesystem"
+    READ = 'read'
+
+COMMANDS = {
+    '/fs': ACTOR_TYPE.FILESYSTEM,
+    '/read': ACTOR_TYPE.READ
+}
+
+@ray.remote
+class Router():
+    """Routes messages to the correct actor. To register 
+    new actors, add the actor type in the `ACTOR_TYPE` 
+    enum and then add a corresponding command in the 
+    `COMMANDS` dictionary.
+    """
+
+    def __init__(self, log):
+        self.log = log
+
+    def route_message(self, message):
+        
+        # assign default actor
+        actor = ray.get_actor(ACTOR_TYPE.DEFAULT)
+
+        if(message.body.startswith("/")):
+            command = message.body.split(' ', 1)[0]
+            if command in COMMANDS.keys():
+                actor = ray.get_actor(COMMANDS[command].value)
+        
+        actor.process_message.remote(message)
+
+
+@ray.remote
+class DocumentIndexActor():
+    def __init__(self, reply_queue: Queue, root_dir: str, log):
+        self.reply_queue = reply_queue
+        self.root_dir = root_dir
+        self.log = log
+        self.index_save_dir = os.path.join(jupyter_data_dir(), '.jupyter_ai_indexes')
+
+        if ChatOpenAINewProvider.auth_strategy.name not in os.environ:
+            return
+        
+        if not os.path.exists(self.index_save_dir):
+            os.makedirs(self.index_save_dir)
+
+        embeddings = OpenAIEmbeddings()
+        try:
+            self.index = FAISS.load_local(self.index_save_dir, embeddings)
+        except Exception as e:
+            self.index = FAISS.from_texts(["This is just starter text for the index"], embeddings)
+
+    def get_index(self):
+        return self.index
+
+    def process_message(self, message: HumanChatMessage):
+        dir_path = message.body.split(' ', 1)[-1]
+        load_path = os.path.join(self.root_dir, dir_path)
+        loader = DirectoryLoader(
+            load_path, 
+            glob="**/*.txt",
```

Comment:
Do we want to expand the set of file extensions to cover code, markdown, rst, latex, etc?